src/metabase/csv.clj

```diff
@@ -0,0 +1,96 @@
+(ns metabase.csv
+  (:require
+   [clojure.data.csv :as csv]
+   [clojure.java.io :as io]
+   [clojure.string :as str]
+   [medley.core :as m]
+   [metabase.search.util :as search-util]
+   [metabase.util :as u]
+   [metabase.util.i18n :refer [trs]]))
+
+;;           text
+;;            |
+;;            |
+;;       varchar_255
+;;            |
+;;            |
+;;          float
+;;            |
+;;            |
+;;           int
+;;            |
+;;            |
+;;         boolean
+(derive ::boolean     ::int)
+(derive ::int         ::float)
+(derive ::float       ::varchar_255)
+(derive ::varchar_255 ::text)
+
+(defn value->type
+  "The most-specific possible type for a given value. Possibilities are:
+    - ::boolean
+    - ::int
+    - ::float
+    - ::varchar_255
+    - ::text
+    - nil, in which case other functions are expected to replace it with ::text as the catch-all type
+
+  NB: There are currently the following gotchas:
+    1. ints/floats are assumed to have commas as separators and periods as decimal points
+    2. 0 and 1 are assumed to be booleans, not ints."
+  [value]
+  (cond
+    (str/blank? value)                                      nil
+    (re-matches #"(?i)true|t|yes|y|1|false|f|no|n|0" value) ::boolean
+    (re-matches #"-?[\d,]+"                          value) ::int
+    (re-matches #"-?[\d,]*\.\d+"                     value) ::float
+    (re-matches #".{1,255}"                          value) ::varchar_255
+    :else                                                   ::text))
+
+(defn- row->types
+  [column-names row]
+  (->> row
+       (map vector column-names)
+       (map (fn [[column-name value]] [column-name (and value (value->type (search-util/normalize value)))]))
+       (into {})))
```

Comment:
Optimization: The `column-names` are in the same order on every row, so we don't need to bring in column-names here. The result of row->types could be a vector of types, not a map from column-name to type.